### ISSUE TITLE
Wrong type when searching from enum in collection in Hibernate 4.3.9+

### ISSUE THREAD

**Issue by zdila** [querydsl/querydsl#1500]
Our class `User` has a property `roles` which is `@Enumerated(EnumType.STRING) private Set<Role> roles;`. When the predicate is `QUser.user.roles.contains(role)` then we are getting the following exception:

```
Caused by: java.lang.IllegalArgumentException: Parameter value [ADMIN] did not match expected type [com.example.Role (n/a)]
	at org.hibernate.jpa.spi.BaseQueryImpl.validateBinding(BaseQueryImpl.java:874) ~[hibernate-entitymanager-4.3.11.Final.jar:4.3.11.Final]
	at org.hibernate.jpa.internal.QueryImpl.access$000(QueryImpl.java:80) ~[hibernate-entitymanager-4.3.11.Final.jar:4.3.11.Final]
	at org.hibernate.jpa.internal.QueryImpl$ParameterRegistrationImpl.bindValue(QueryImpl.java:248) ~[hibernate-entitymanager-4.3.11.Final.jar:4.3.11.Final]
	at org.hibernate.jpa.internal.QueryImpl$JpaPositionalParameterRegistrationImpl.bindValue(QueryImpl.java:337) ~[hibernate-entitymanager-4.3.11.Final.jar:4.3.11.Final]
	at org.hibernate.jpa.spi.BaseQueryImpl.setParameter(BaseQueryImpl.java:674) ~[hibernate-entitymanager-4.3.11.Final.jar:4.3.11.Final]
	at org.hibernate.jpa.spi.AbstractQueryImpl.setParameter(AbstractQueryImpl.java:198) ~[hibernate-entitymanager-4.3.11.Final.jar:4.3.11.Final]
	at org.hibernate.jpa.spi.AbstractQueryImpl.setParameter(AbstractQueryImpl.java:49) ~[hibernate-entitymanager-4.3.11.Final.jar:4.3.11.Final]
	at com.mysema.query.jpa.impl.JPAUtil.setConstants(JPAUtil.java:55) ~[querydsl-jpa-3.6.6.jar:?]
	at com.mysema.query.jpa.impl.AbstractJPAQuery.createQuery(AbstractJPAQuery.java:130) ~[querydsl-jpa-3.6.6.jar:?]
	at com.mysema.query.jpa.impl.AbstractJPAQuery.count(AbstractJPAQuery.java:81) ~[querydsl-jpa-3.6.6.jar:?]
	at org.springframework.data.jpa.repository.support.QueryDslJpaRepository.findAll(QueryDslJpaRepository.java:141) ~[spring-data-jpa-1.8.2.RELEASE.jar:?]
...
```

For the predicates having only simple equality test with the enum it works. Also it works with Hibernate 4.3.8.Final and before.

The problem is probably at `com.mysema.query.jpa.JPQLSerializer.visitAnyInPath(Class<?>, Operator<?>, List<? extends Expression<?>>)`. Instead of `args = ImmutableList.of(ConstantImpl.create(constant.name()), args.get(1));` there should probably be `args = ImmutableList.of(ConstantImpl.create(constant), args.get(1));`

But maybe the problem is in the Hibernate and they've broken the compatibility. Could you please check it?

**Comment by timowest** [comment]
The reason why the String binding was used is https://hibernate.atlassian.net/browse/HHH-5159

**Comment by martinsefcik** [comment]
@timowest : Are you planing to fix it in 3.6.x branch ?

**Comment by johnktims** [comment]
@martinsefcik, this has been backported and will be released in 3.6.9 on October 30th.

**Comment by martinsefcik** [comment]
Thanks!
